### ISSUE TITLE
'ask for' a sprite variable of a clone never terminates

### ISSUE THREAD

**Issue by ToonTalk** [jmoenig/Snap#2242]
![image](https://user-images.githubusercontent.com/2138696/47786526-ef0c4800-dd03-11e8-8d67-09d2b72ed6c0.png)

'location of clone' works fine but

'ask clone for location' doesn't terminate

**Comment by joker314** [comment]
This totally should always terminate, but could you try putting the 'location' getter in a ring when asking for it?

**Comment by ToonTalk** [comment]
Doing so causes the ask to return 0.

**Comment by ToonTalk** [comment]
But it works fine to 

'ask clone for x position'

What is the difference?

**Comment by jguille2** [comment]
This is the difference between a variable and a reporter (they have the same shape, but they are different).

_x position_ is a reporter and the _ask_ block  uses its definition.

With variables it is different. We have local and global variables. We see their names in "their reporter", but inside them, it has a context (I don't know if "context" is the right name... but it is the idea).

**Comment by jguille2** [comment]
I amend a little my explanation...

The problem is not between _reporters_ and _variables_. The question is more about _local_ and _global_ scopes. The problem is the same with _methods_ (reporters or commands).

You can see a long discussion of #1888 and this issue is shown.

**Comment by ToonTalk** [comment]
But if the issue is about local scope then why again does 'x position' work while my own local sprite variable doesn't? Both are conceptually local to the sprite.

**Comment by jguille2** [comment]
No
 _x position_ is a global method. The result of this method is different for each sprite, but the "method" is the same.

But as I said, this is not a simple question. We like our simple and intuitive environment... but some simplifications have their weak points.

I try to explain these "hidden context of the blocks"...

## 1- Blocks definition has their own context
An example. I make a global variable ("global") as a list with two reporters. Both reporters are apparently the same ("x position") but not really: they are reporters of different sprites (I added them to the list from different sprite script area). Then you can see...

![globaxpositions](https://user-images.githubusercontent.com/2926315/47822354-15fa6680-dd64-11e8-9ead-c4469076cfcf.png)

The "same" reporter reports different values. And I am in the same (really the same) sprite. Why? Methods (reporters and commands) has a defined context.

## 2- Ask and Tell blocks change the context of the **global** blocks

Using the same "global" variable from previous example, we can see...

![globalxpositions2](https://user-images.githubusercontent.com/2926315/47822784-b3a26580-dd65-11e8-90af-0c6486a9bd5f.png)

Then, here (with _ask_ and _tell_) we use the method (is the same), and the context is defined outside. Here the result is the same.

## 3- Different case (but same behavior) of local blocks

Using local blocks (variables, commands, reporters...) the things are not the same. Local blocks can take the same name in different sprites... but their definition may be quite different. So, the problem is not only the context ("who are executing") ... the problem is "what is this block doing"

We discussed this in the past. This is not white or black... but current implementation  changes the context and also the definition of the method.

So, _ask_ and _tell_ is using basically the name of the method. Maybe we can say that polymorphism has been preferred to the first class behavior (ony here!).

## 4- Different case (and different behavior) of local variables 

I've remake the first example. Now "global" has two "a" variables. Both are local, from different sprites.

![localvars](https://user-images.githubusercontent.com/2926315/47823618-5d372600-dd69-11e8-9e6a-d158b0173fd0.png)

Here, the "hidden" context is running ok (like 1st case), but _ask_ and _tell_ does not change the context. Why? Because here (when the "reporter" is a variable) definition and context are tied. If we are talking about the "locar variable _a_ from the sprite _Sprite1_", the Sprite1 is the definition and the context.

---
And I stop the roll :newspaper_roll: I hope it is useful :roll_eyes: :roll_eyes: 

Joan

**Comment by ToonTalk** [comment]
Perhaps this is a different issue but I don't really understand why the first one doesn't work (and it is unclear what values are in the list it returns -- they aren't numbers).

![image](https://user-images.githubusercontent.com/2138696/47850698-10396b00-ddce-11e8-848f-c416d5019025.png)

**Comment by joker314** [comment]
Try placing a grey ring around the "of" block in the first example.

_Snap!_ will evaluate the arguments of "map", and _then_ invoke "map". This means that what "map" sees is something like:

```
map (2) over [Alice, Bob, Charlie, ...]
```

But what you want is to pass in the _block itself_. To say "These blocks are the data itself; their result isn't" we put a ring around them (like has been done in the second example).

Tip: you can right click the block and hit "ringify" iirc.

**Comment by ToonTalk** [comment]
Right. But what is so confusing is that map already has a ring and dropping the 'of' reporter removes it but not when dropping the 'ask' reporter. 

Also what should `map (2) over [Alice, Bob, Charlie, ...]` do? What it returns now is mysterious.

**Comment by joker314** [comment]
I don't know why the ring is removed when you drop in that block. I'm not too experienced in Snap<em>!</em>... sorry!

However, under the hood, "map" uses "call". Now, with 2 it errors (correctly):
![Calling 2 gets an error](https://user-images.githubusercontent.com/18113170/47851882-dd917180-ddd1-11e8-8c75-09eb85948d05.png)

However, I was mistaken. You don't get a number when you evaluate that block. You get an empty string.

![Calling the empty string yields an error-free empty string return value](https://user-images.githubusercontent.com/18113170/47851933-07e32f00-ddd2-11e8-9deb-057e59762716.png)

I think this behaviour is unintuitive, but there might be a good reason why it exists.

**Comment by kephas** [comment]
@jguille2, your explanation is pretty clear, but I'm not sure I'll remember it easily. Which doesn't bode well for how kids would deal with it.

**Comment by brianharvey** [comment]
In my opinion, there should be no difference in behavior between the primitive ASK/TELL and the one you can define (we did define in Tools pre-primitive-ones) using OF.  The difference described above is a bug.

Which is the correct behavior?  I'm less certain about this, but I'm _pretty_ sure that for pedagogic reasons there should be no difference between variable getters and other reporters.  (I don't subscribe to the "variable blobs aren't reporters" theory, although I think Jens does.)  So, OF applied to a local variable getter should return a getter for the corresponding-name variable of the other sprite, and therefore, so should ASK.

As for the disappearing ring with OF inside CALL, this is a semantic mess we brought upon ourselves by repurposing the Scratch OF block for methods and sprite variables.  (I have just this minute decided we should retire the phrase "sprite-local variables" because of the conflicting use of "local" for script variables.)  In its original use, the OF block never returned a procedure.  But our extended version _does_ return a procedure when called with a procedure in its left input slot, and basically you always want to call that returned procedure, not call OF itself, in a CALL block.  So we remove the ring, just as we do for variable getter blocks and for ITEM.  The correct behavior would be to remove the ring if and only if the left input is a (picture of a) block, rather than one of the built-in attributes.  But I doubt Jens will be willing to do that, especially since many people will first drag the OF block into the CALL block (with its default left input of "costume #"),  and _then_ change the left input to a method, which would require that the OF block notice that its continuation is a CALL block, and erase the ring.  So maybe the next-best thing would be not to erase the ring at all; the user would have to do it explicitly.

I am labelling this a bug because of the original issue, in which ASK behaves differently from OF.

**Comment by jmoenig** [comment]
yep, looks like a bug to me

**Comment by joker314** [comment]
This isn't the original issue, but I don't understand why `call`ing an empty string doesn't error? Is it useful in recursion or something? Otherwise, it would probably be very useful in catching mistakes of students?

**Comment by jmoenig** [comment]
let's keep it on topic, then :-)

**Comment by jguille2** [comment]
Hi,
I know these long threads mixing questions, mistakes, doubts... and also design considerations can be unprofitable... but I try to continue.

I want to write down a small summary  of the current behavior of all this (tell/ask, local methods and variables...) that can be useful (not only for design considerations... also to present to teachers, students...) but first, I need a clear answer to one of the aspects that I have already explained in these threads.

I write my question with an example to be clearer. I don't know if this current behavior is right or not (@jmoenig , @brianharvey  can answer me), but  I think _rings_ must  forget their context to achieve the best implementation and to have the optimal combination of late-binding (i.e. polymorphic) vs. early-binding (i.e. closurized).

-----
My question into an example...

* I have a global variable ("global") and, from Sprite1, I set this variable to a global method

![gobalassigned](https://user-images.githubusercontent.com/2926315/47995100-8ef71680-e0f4-11e8-8295-3ac1195c01ee.png)

* Then, when I run this method into other context (Sprite2)...

![runglobal](https://user-images.githubusercontent.com/2926315/47995163-b817a700-e0f4-11e8-8251-59b326460abb.png)

* ...the result is that the method is applied to the original context. I run the block into Sprite2 scripting area, but the movement is applied to Sprite1.

**Is this right? I expected that rings had no context**, and the result context would be the receiver sprite.

* Yes I know that I can do

![tellglobal](https://user-images.githubusercontent.com/2926315/47995415-8d7a1e00-e0f5-11e8-98de-961293b95adf.png)

and I can explain how to pass local methods between objects (using variables with our first-class features, "closurizing"). But it would be more clear if I use "tell/ask" only to call other sprites and "run/call" when current sprite is the receiver.

-----
I have another idea to use "closures" in a fast way... but I don't want to complicate this... and I wait for your answers.

Thanks,

Joan

**Comment by brianharvey** [comment]
Jens and I were talking about this issue yesterday (great having him here!) and I think I've convinced him that users should be able to get whichever behavior they want, and that it makes most sense if TELL SPRITE1 TO ... should make sprite1 the receiver, and therefore we should make "... OF SPRITE1" give us a receiver-less method.  In other words, methods normally imply a receiver, and it's OF that disrupts the normal functioning of the universe (sorry, I'm being influenced by today's Girl Genius...) by stripping the receiver from the method.  In your example there's no OF, and so the method retains its implied receiver.  So that's correct behavior.

**Comment by jguille2** [comment]
Hi,
I come back to this issue. My aim is to clarify this question. We don't need quick decisions (I think the original issue can be closed), but we have different threads with questions about scopes, global/local blocks... and I think it's a good time to a design discussion.

I try to explain all of this in the next three posts:

1. Current behavior: Rings storing their context. Considerations and possibilities.

2. Possible change to have rings without context.

3. Running local methods in other sprites.

**Comment by jguille2** [comment]
## 2. Possible change to have rings without context.

* ![runglobal](https://user-images.githubusercontent.com/2926315/48128919-440e0800-e288-11e8-92c8-75b8626a82e7.png)

  Here, RUN/CALL can be the "main methods". When somebody "RUN something" the action runs where the block is. We don't need to change nothing (order of the blocks nor adding more options). TELL/ASK will be used only to call other sprites. The block "CONTEXT OF ring" has no sense... because rings wouldn't have context.

* Storing contexts.
  But if somebody wants to go further and store methods and contexts, he can do it.
![setglobalwithcontext](https://user-images.githubusercontent.com/2926315/48128982-70c21f80-e288-11e8-8b70-cb819994dd1e.png)
  and then
  
  ![globalreciverandmethod](https://user-images.githubusercontent.com/2926315/48129017-8a636700-e288-11e8-9c3d-74b803247b35.png)

  And the method can be used (with or without the receiver)

  - ![tellitemsofglobal](https://user-images.githubusercontent.com/2926315/48129039-a49d4500-e288-11e8-8efe-15e50a823389.png)
     will be executed in the original sprite, but
  - ![runglobal](https://user-images.githubusercontent.com/2926315/48129050-b1ba3400-e288-11e8-9887-40d1eeccf699.png)
    will be executed in the current sprite.

* I think this option is clearer for simple behaviors, and if somebody wants a more complex implementation, he must do more complex actions.

**Comment by jguille2** [comment]
## 3. Running local methods in other sprites

Regardless of the option (rings storing context or not), the "problem" of calling a local method from other sprite (to use it for itself) is the same.

Current behavior (and I think it's fine) use polymorphism and then: If I want to run a local method "sing" of other sprite, I will run my "sing"method (not its method) and, if I haven't this method, I can't run it. The OF block does not do it either.

But, can I "teach" other sprite to do something? Yes, using our first class blocks! See it into an example:

- We local variables "local method" in Sprite1 and Sprite2
  
  ![localmethod1](https://user-images.githubusercontent.com/2926315/48129592-45403480-e28a-11e8-853d-78224f904e13.png)
  and
  ![localmethod2](https://user-images.githubusercontent.com/2926315/48129613-538e5080-e28a-11e8-9247-1fa116a49046.png)

- Then, from Sprite2 I have
  
  ![reportlocalmethod2](https://user-images.githubusercontent.com/2926315/48129654-728ce280-e28a-11e8-822b-a1a083fc0372.png)
  and
  ![reportlocalmethodof1](https://user-images.githubusercontent.com/2926315/48129963-65242800-e28b-11e8-8042-f807bfbb7ad6.png)

- And now I can use "local method" of Sprite1 inside Sprite2

  - in current implementation
  ![telllocalmethod](https://user-images.githubusercontent.com/2926315/48129742-ab2cbc00-e28a-11e8-955c-ac3954d627b5.png)
  - and with the modification of "rings without context" will be
  ![runlocalmethod](https://user-images.githubusercontent.com/2926315/48129766-c5ff3080-e28a-11e8-9df2-2ce9730353e6.png)

----
Yes, we can do it... but maybe it's annoying... because we can not use our custom Blocks to do it.

How can I do it with custom blocks? I need a reporter that reports the content of the block (a normal closure issue)
Then, if I will have something like

![contentoftest](https://user-images.githubusercontent.com/2926315/48129811-e8914980-e28a-11e8-8a57-1a8f10bfd953.png)

With this, I could do the same as I done with local variables, calling the "contents" of the blocks.

And yes, I know that this is no so easy! Local blocks can have other local blocks, the block itself (recursivity), local variables... and their own input variables. This is only a draft idea. In fact, in the heart of this, there is the differentiation between variables and reporters.

**Comment by ToonTalk** [comment]
Note that the `ask `doesn't loop but `think `of it does.

**Comment by jguille2** [comment]
Hi @ToonTalk,

But I thought that my first response had answered this.

1. In one hand, as @joker314 had said, the reporter must be ringified to run. This is general because ASK needs a ring. You can test with other reporters and you'll get the same issue.

2. As you said, with the ring it reports 0 (and not 5). And this is because "location" (this reporter you are dragging from the palette) is the "location of the sprite" (and not the "location of the clone"). Then, you need "location OF clone"

  As we have seen in previous posts, you must use:
  ![issue](https://user-images.githubusercontent.com/2926315/48142625-649c8900-e2ad-11e8-9778-cc0c3de711ec.png)
  But, in fact, this is the same that the direct "location OF clone".

**Comment by ToonTalk** [comment]
Yes I understand that ringifying addresses the problem but should Snap! go into an infinite loop because one does a `think `of an improper `ask`?

**Comment by jguille2** [comment]
Ok,

You are right,

Reporters always report something (sometimes an empty string when they can not evaluate nothing) or an error. The behavior will not be more different (maybe the "white loop" will be a "red loop" of an error).

But maybe this must be amended.

Thanks,

Joan

**Comment by brianharvey** [comment]
I agree that the original bug is a bug and should be fixed.  RUN and CALL should check that their input is a command or a reporter, respectively.

As for the design issue, I agree with Joan that a ringed script that contains an invisible, implicit receiver is confusing.  My proposal solves this problem by eliminating that possibility.  The proposal is that ringifying a block or script reports a pure, receiverless block or script, even though the blocks run without ringification do have implicit receivers.  (Maybe even that could be changed, but that would be a big change all over the code, whereas this proposal just changes the ringification process.)  When a sprite invokes a ringified block or script, it uses itself as the receiver.  Ringified procedures reported by OF are no different; they do not remember the sprite that originated them.

If you RUN or CALL a procedure, it runs with you as receiver.  If you TELL or ASK another sprite to invoke it, then it runs with that other sprite as receiver.

I believe that the current arcane business about hidden receivers is just a design error inherited from Scratch.

**Comment by brianharvey** [comment]
Actually, looking back, that's _not_ what I proposed before.  But it's what I'm proposing now; there's nothing special about OF.  Ringification is what's special.

**Comment by jmoenig** [comment]
Snap's implicit receiver, captured in an environment, is one of the design decisions I'm most proud about. Not a chance I'll ever change that. If you think it through thoroughly I'm sure you'll agree.

That said, I understand that some of you want to get the "pure" function and run it in some other environment, even if that doesn't make much sense, e.g. if things like variable references don't exist in the target environment. I agree that we will probably explore such things with macros, BUT the goal (my goal, at least) is to enable introspection and debuggin, NOT EVER to let you write programs that zombiefy other objects. You know, I do come out of Smalltalk and OOP, and the maxim there is that the receiver decides whether and how to respond to requests and messages, never the sender.

So, yeah, if Snap! hangs that's a bug and I'll address it. When you talk about "pure" functions in the context of a sprite-local method you're full of crap :-)

**Comment by brianharvey** [comment]
I wasn't thinking of named functions declared sprite-local so much as anonymous ringed scripts.  In those cases, there's nothing _obviously_ sprite-local about them.  If you're a user thinking about the primitive MOVE block, I claim the obvious way to think about it is that MOVE is a global procedure, and whichever sprite invokes it is the one that moves.  What I'm calling a design error is that, there's an _invisible, implicit_ receiver built into each sprite's MOVE block.  If sprite1 does TELL sprite3 TO [MOVE 100 STEPS], even though you dragged the MOVE block out of sprite1's palette, the clear intent is that it's sprite3 that should move.

In fact, never mind even ASK, TELL, OF, and rings.  If I construct a script in sprite1's scripting area that includes a MOVE block, and then I drag the script over sprite3's icon in the sprite corral, a copy of that script appears in sprite3's scripting area -- except that you've done handsprings to convert the MOVE-of-sprite1 block into a MOVE-of-sprite3 block, since that's _clearly_ what the user wants.

It's those MOVE and TURN blocks inside a ringed script that I want to be "pure."  The straightforward way to achieve that would be just to make them receiverless global blocks in the first place, but changing Snap<i>!</i> to do that would, I'm sure, be a huge undertaking.  That's why I'm suggesting that ringifying a MOVE (or a larger script containing a MOVE) should strip the receiver from the block, keeping all the hassle in one place in the code.

As for named sprite-local methods, I agree with you that each sprite should run its own method in response to a message from another object.  I believe that that's what my proposal implies.  If I construct an instruction TELL sprite3 TO [method] in sprite1, by dragging sprite1's METHOD block into the TELL block, sprite3 should nevertheless run its own METHOD.

But what if, for some reason, sprite1 _wants_ to run sprite3's METHOD?  Well, first we have to be clear about what that entails.  To be specific, let's say sprite1's METHOD is MOVE 100 STEPS, while sprite3's METHOD is MOVE 300 STEPS.  What should it mean if sprite1 says RUN [[METHOD] OF sprite3]?  I claim it should mean that sprite1 (not sprite3!) should move 300 steps.  We want the body of sprite3's method, applied to sprite1 as receiver.

I think all this avoids zombification, if I'm correctly understanding what you mean by that.

As we discussed the other day, there's one obvious situation in which a user would want this mechanism, namely, sprite3 is the parent of sprite1, and sprite1's METHOD wants to include its parent's METHOD.  In that case it's clear that sprite1 should carry out the action specified by sprite3's method.

**Comment by jmoenig** [comment]
I disagree about "pure global" functions. That would be - if at all - a special if you *only* ringified *one single* ```move 100 steps``` command. It's quickly turning into a different thing once you ringify more complex expressions or statements (scripts). Snap is so full of local state and Snap's sprite's are so extensible with attributes, parts, events and methods that there's nothing "pure" in there at all. That's also why one way to make a local method is to have a sprite assign a script to a shared variable (regardless of sprite-local or global) and allow another sprite to ```run/call``` it.

Calling a ```super``` method might be special, but if I'm going to do it at all, it will probably be a special case, like in real OO languages, not a zombie-handle. But - you know - I'm not even sure about ```super``` any more. Yeah, to be honest, I rarely used it in Smalltalk, and I rarely ever use it any more in Snap's own source code. I get it that we might want to teach about it, but I'm not at all sure about whether it's such a great and beneficial feature at all. It's certainly not something I can't live without.

That said, yes we'll make it so we'll get introspection and debugging, but I'd like to take special precautions to actively prevent programs that use one sprite to remote-control another. Because *that's* - to me - the most, most, utmost - gospel of OOP :-)

**Comment by brianharvey** [comment]
Jens, sorry, I don't get it.  I think what I'm proposing does what you want -- preventing one sprite from making another sprite run the first sprite's method -- in a much simpler, more comprehensible way.

Oh, wait, I just had an idea, not about what we should do, but about why we're arguing.  I see a block in a script as being akin to a procedure _call,_ agnostic as to what that procedure actually does.  You, I think, see a block as akin to a procedure _definition,_ including the procedure's meaning.  Is that right?

So, let's say sprite3 has its own local method for MOVE.  If sprite1 TELLs sprite3 to MOVE, then sprite3 should do what MOVE means to it -- its own local method.  When I say that the MOVE block should be "pure," I mean that it's a pure procedure _call,_ calling whatever procedure that block calls for that sprite.

The situation you're trying to prevent is that sprite1 tells sprite3 to run sprite1's method.  And we are in agreement about that.

The situation that I want to _allow_ is that sprite1 wants to run sprite3's method -- run it in sprite1.  I agree with you that this is probably rare.  But that's why my understanding of OF would allow, not the opposite direction.

If I'm still misunderstanding you, please give an example of the "more complex expressions or statements."

**Comment by jmoenig** [comment]
Hmm... I thin you probably did understand me right. We agree that sprite1 should not be able to make sprite2 run sprite1's stuff. 

For me it goes further than that, and here's where I think we might disagree:

sprite1 should also not be able to run sprite2's local stuff, even if it so wishes, because stuff inside objects are all private.

And even further yet:

"private stuff" is not just local methods, but also any script ringified by a sprite, because whatever happens inside a scripting area is *local* to the sprite owning that scripting area.

Whenever we ask/tell another sprite to run/call something, and - likewise - whenever a sprite requests something local "OF" another sprite, - by OOP design - it never gets any information about the other sprite, and can only look at what it gets in its own context (in the case of "OF") or in the context of the other sprite (in the case of ASK/TELL).

That, to me, is the evangelism of modularization. It avoids tightly coupling sprites to each other.

Now, where I might soften is when things stay "in the family", i.e. inside inheritance chains. That's the question of "super" methods. Also, where I might soften is the case of introspection for debugging. But, since the latter can be abused to write zombie-remote controllers I want to think about ways to prevent that. One thing I can think of right away is that none of those informations to be gleaned by introspection will be permanent, they're all going to be transient to that particular session.

But it's still too early for such design. Right now the rule of OOP is that no sprite can run another sprite's local stuff.

**Comment by brianharvey** [comment]
>"private stuff" is not just local methods, but also any script ringified by a sprite, because whatever
> happens inside a scripting area is local to the sprite owning that scripting area.

Oh, no, if you're that strict about it, there's no way to pass another sprite a message at all!

I think this is in part about a difference -- not better, not worse, but different -- between blocks languages and text language.  In a text language, it's easy to say that a message is a piece of text, that the other object interprets however it wants.  But I think you'll agree that we don't want users to have to type the name of a block (or the blockspec, even worse, because title text doesn't all come before the inputs, so there's really no "name") to TELL or ASK; we want them to use ringified blocks.  And if sprite1 says TELL sprite3 [[MOVE 10 STEPS]], we want sprite3 to move, not sprite1.

I guess I'm saying that in moving from non-OOP Snap<i>!</i> 4.0 to OOP Snap<i>!</i> 4.1ff, we shouldn't make anything _harder_ for the user.  If I can live with having commands in Snap<i>!</i>, even though it's supposed to be Scheme and emphasize functional programming, you can live with un-owned ringed scripts, even though it's supposed to be Smalltalk and emphasize OOP.

Putting it another way, in addition to being a Scheme and a Smalltalk, we're also a Scratch, making easy things easy for beginners who haven't pledged a paradigm fraternity.

The OOP modularity and hiding rules are for teams of 5000 programmers, not for students, who want to be able to examine everything all the time.  You have better things to do with your time than "prevent[ing] the latter [being] abused to write zombie-remote controllers."  That's like me saying we should find a way to prevent SET being used to change the value of an existing variable, to force functional programming.

**Comment by brianharvey** [comment]
P.S.  Your proposal would break a ton of existing projects.  For example, in BJC there's an exercise very early on about two sprites telling a knock-knock joke:

"Knock knock."
          "Who's there?"
"Sam."
          "Sam who?"
"Sam and Janet Evening."

And the way we do it is

SAY [Knock knock.] FOR 2 SECS.
TELL (other sprite) [SAY [Who's there?] FOR 2 SECS.]
SAY [Sam.] FOR 2 SECS.
TELL (other sprite) [SAY [Sam who?] FOR 2 SECS.]
SAY [Sam and Janet Evening.] FOR 2 SECS.

That's just way easier than using five different named messages!

And, by the way, this reminds me of a case in which we need min/default/max: the TELL block needs a rightarrow which, when clicked, exposes the text AND WAIT.

**Comment by brianharvey** [comment]
P.P.S.  [Not just editing the original message because Github isn't letting me, as punishment for running a browser it doesn't approve of.]  I don't even think you're right about what OOP is about!  It's a metaphor for multiple independent agents inside the computer.  Sprite-local variables exist so that different objects can use the same _name_ to mean two different things -- partly to let a global procedure work on any sprite.  But our sprites aren't _adversaries!_  They are a convenience for the one programmer, and they cooperate to tell the programmer's story -- even if, in that story, they _portray_ adversaries.  They are actors, just like the real actors in a play, who cooperate to tell the story even if their characters are enemies.

**Comment by brianharvey** [comment]
Okay, clearly I'm not understanding you.  I'm sorry.  I'm going to go to sleep now, and tomorrow we can talk about this privately.

**Comment by jmoenig** [comment]
Okay, the original bug report is about an operation that "never terminates". This isn't really a problem, because it is user-stoppable by clicking on the red stop sign, and does not "hang" or "crash" Snap. Also, it only doesn't terminate if the value of the variable is numeric, because it is trying to look up a variable with the variable's value. The semantics of having to ringify a variable to look it up using ASK are correct, and there is also an alternative way of directly looking up a local variable using the OF block. No need to for all this complicated passing of anonymous functions just to read a value. I'm going to disagree about this even being a bug and remove the "bug" label. This is a question which has been answered, followed by an ongoing - interesting and lively - discussion about esoteric programming practices :-)

**Comment by ToonTalk** [comment]
Regarding the bug label the code uses up 100% of the CPU until you stop it. I call that a bug. Especially since the user may not know that its happening since as you say everything continues to work as normal.

Main discussion here is very interesting...

**Comment by jmoenig** [comment]
Oh, but I'm not talking about sending messages to particular sprites, just the plain old Scratch-style broadcasts. Many BJC projects I see don't ever use any events but end up CALLing and RUNning everything explicitly. I'm only pointing out in response to Brian's "we're also Scratch" that there's often nothing wrong with triggering events and letting sprites respond to them whichever way they like. Again, to me the issue is decoupling sprites and modularizing projects. For this events and encapsulated scripts are awesome, whereas RUN and CALL across sprites is suboptimal.

**Comment by ToonTalk** [comment]
I'm also a big fan of encapsulation. It is not only for projects with 5000 programmers. Modularity makes it easy to reason about programs And as you mention combining projects. But I'm confused about `of `. Doesn't it violate encapsulation?

**Comment by jmoenig** [comment]
yes, I agree. I've argued vehemently against ```of``` when we introduced it to Scratch (I think it was v1.3). John said it was okay for him because it would only let one sprite *get* a value of another sprite's local var or attribute, but not *set* them. I still believe it's an original sin, but have come to live with it, especially since there is no way to pass a parameter otherwise in Scratch. That's why I'm adamant about ```of``` only ever returning something in the context of the *receiver*, (same with TELL and ASK), and never to be able to force-remote-control (zombify) another sprite. With *this* interpretation it's easier for me to overlook the sheer ugliness of directly querying into another object's internals.

**Comment by jguille2** [comment]
Uff!

Perhaps it is healthier to continue the discussion later, outside this thread...

As I understand it, there are two things:

* The origin of this issue. Maybe not a "bug"... but maybe yes
  - All the reporters are highlighted before finishing. And in the end (if it arrives) always report something (a value, an empty string... or an error).
  - But ASK is not doing this in the reported example. It finishes its action and report nothing. Then, ASK is not highlighted (it has already finished) and SAY (it could be any other block) is waiting for ever.
  - If this happened in a custom block, an error will occur, saying "Reporter didn't report".
  - Following this current Snap! behavior, maybe ASK must report an error in that case. I don't like these errors messages very much... but I am not thinking in errors, I'm thinking about "the flow" of the script.

* Design discussions
  - I don't want to continue a "global discussion" about "almost everything".
  - My doubt about "receiver stored inside the rings" was concrete. I believe I have understood that the decision is clearly to keep this behavior. For me it is fine. I will try to document and explain clearly this to our users.
  - I don't know what do you think about my proposals to help to explain it to users. Basically two things:

    - To add a "myself" to the dropdown options for ASK and TELL and to put these blocks upper than RUN and CALL. The idea is students use first these blocks because their behavior is clear (no problem with hidden recivers), and RUN and CALL for advanced users.
    - The idea of the block "Context of (ring)" can helps to explain the behavior or rings. I could be a primitive beside RUN/CALL, or inside a library with tools for OOP... The idea is it helps to explain the more complex idea of "context/receivers of a ring".
     - I don't want to reopen the ideas about "Content of a ring" because this is a more complex and global thread (about scopes, private stuff, remote-controling...)

And that's all!

Joan

**Comment by brianharvey** [comment]
I don't want to argue about the semantics of the word "bug," but it's suboptimal for a user who makes a mistake to be confronted with an infinite loop rather than an error message.  This happened to an actual user, Ken, who's an OOP expert and still did the wrong thing here.

If I'm understanding the problem correctly, its origin is our policy of removing the grey ring when a variable is placed in the function slot.  That policy does what the user meant 99% of the time, and I'm not proposing to change it, but I do have some sympathy for a user who is tripped up by it the other 1% of the time.  If I'm understanding that correctly, I would like to see a _very specific_ error message, "The variable you gave as input to ASK doesn't have a block or script as its value.  If you meant to ASK for the value of the variable itself, you must use the 'ringify' option to specify that."  Or something.  (I'd like to see very specific error messages in general.  "You are trying to ALL BUT FIRST OF an empty list.  Did you forget the base case of a recursion?")

**Comment by ToonTalk** [comment]
I agree. And in this case the infinite loop was within `say `which is very different from a user's infinite loop.

I don't really understand why `say` is doing this. When I opened Chrome's task manager and then clicked on the block the CPU usage averaged 100% until I stopped it. If it is just busy waiting then why isn't it a lower percentage?
